### DISCUSSION TITLE
How can I add a little space to the last page?

### DISCUSSION THREAD

**Discussion by TarasBounty** [espresso3389/pdfrx#131] [Q&A] [answered]
I have implemented my own controls placed on top of the pdfrx widget.

`Stack(
  children: [
    _PdfContent(),
    Positioned.fill(
      bottom: 20,
      child: Align(
        alignment: Alignment.bottomCenter,
        child: _Controls(),
      ),
    )
  ],
)`

The problem is that on the last page my controls are overlapping the PDF content. Is it possible to add a padding on the last page only?

@espresso3389

**Comment by TarasBounty** [comment]
Or not space, any widget.Using current version it's only possible to display any widget over the page, but I want smith like Column() widget behaviour or padding to page.

**Comment by espresso3389** [comment]
For extra paddings, you can use [PdfViewerParams.layoutPages](https://pub.dev/documentation/pdfrx/latest/pdfrx/PdfViewerParams/layoutPages.html) to add extra paddings on the bottom:

```dart
PdfViewerParams(
  layoutPages: (pages, params) {
    final width =
          pages.fold(0.0, (w, p) => max(w, p.width)) + params.margin * 2;
  
    final pageLayout = <Rect>[];
    var y = params.margin;
    for (int i = 0; i < pages.length; i++) {
      final page = pages[i];
      final rect =
          Rect.fromLTWH((width - page.width) / 2, y, page.width, page.height);
      pageLayout.add(rect);
      y += page.height + params.margin;
    }

    // extra padding on the bottom
    const extra = 100.0;
    return PdfPageLayout(
      pageLayouts: pageLayout,
      documentSize: Size(width, y + extra),
    );
  },
),
```

**Comment by espresso3389** [comment]
The code above does not take the zoom into consideration.
If you also reflect the zoom ratio, you can use controller to obtain the zoom ratio.

**Reply by TarasBounty** [reply]
yeah, I've tried already by myself code like yours above, but it padding look different on mobile or web small screen size or big screen size. Maybe you are right and this is because of zoom... I will try to investigate..

**Reply by espresso3389** [reply]
Typically, to normalize the size, length in layout function, divide them by zoom ratio. It results in consistent looking size.

**Reply by TarasBounty** [reply]
it should be done in `layoutPages` function? Zoom info I can get only from controller and only when controller ready

**Reply by espresso3389** [reply]
`layoutPages` is called everytime zoom is changed so if you could not access controller, you don't have to handle zoom.

**Reply by TarasBounty** [reply]
I don't need to recalculate it everytime, I need only initial zoom. If I've tried to get initial zoom, controller returns `Error: Unexpected null value.`

**Reply by espresso3389** [reply]
Even though you don't need calculation on every zoom change, you should do that. It is needed to keep the internal consistency.

```dart
PdfViewerParams(
  layoutPages: (pages, params) {
    final controller = ... // from somewhere
    final zoom = controller.isReady ? controller.zoom : 1.0;
    final width =
          pages.fold(0.0, (w, p) => max(w, p.width)) + params.margin * 2;
  
    final pageLayout = <Rect>[];
    var y = params.margin;
    for (int i = 0; i < pages.length; i++) {
      final page = pages[i];
      final rect =
          Rect.fromLTWH((width - page.width) / 2, y, page.width, page.height);
      pageLayout.add(rect);
      y += page.height + params.margin;
    }

    // extra padding on the bottom
    const extra = 100.0;
    return PdfPageLayout(
      pageLayouts: pageLayout,
      documentSize: Size(width, y + extra / zoom),
    );
  },
),
```

**Reply by TarasBounty** [reply]
idk what I did wrong, but `layoutPages` is **not** called everytime I change zoom via `.zoomUp() /.zoomDown() `

Also, it called only one time when page rendering and `final zoom =  controller.isReady ? controller.currentZoom : 1.0;`  ` controller.isReady` alsways return false, so `zoom ` is always 1 but actually not =(

**Reply by espresso3389** [reply]
Woops, I'm wrong... :(
OK, so layoutPages does not help you.
I should think more about another mechanism to support your case.

**Reply by TarasBounty** [reply]
@espresso3389 any ideas?

**Comment by espresso3389** [comment]
The blocker behind the issue is that the room on the bottom is affected by zoom ratio.
If we simply place some widget or blank space around the pages, it is zoomed.

In my understanding, you want to place some fixed size margin on the bottom of the last page. Is it correct?

**Reply by TarasBounty** [reply]
Yes, it is

**Reply by TarasBounty** [reply]
Do you have an idea how to deal with it? I still need it ;-)

**Reply by TarasBounty** [reply]
I still need it..

**Reply by espresso3389** [reply]
Yes, I understood that!

**Comment by espresso3389** [comment, marked as answer]
@TarasBounty 
1.0.70 contains an update for the issue that calls `layoutPages` for every zoom change. Please evaluate it and give me any feedback.

**Reply by TarasBounty** [reply]
**Everything works perfectly now, thanks!**
One note: previously (on v 1.0.65) my pdf content fills all available space on Web, with 1.0.70 it now renders with grey margins left and right.
